rocPyJpegDecode Python lib & API

essamROCm opened the pull request:
This PR is ready for review. It reflects/implement all what the CV team has discussed in the design reviews.
It has been tested thoroughly, and it passes all CI checks successfully.

essamROCm committed:
Adding rocJPEG to rocPyDecode

essamROCm committed:
Preparing source files

essamROCm committed:
Adding PY paramaters and API interface

essamROCm committed:
Adding enum structures types to PyBind11 API

essamROCm committed:
Adding all the rocJPEG APIs with PY sample startup code

essamROCm committed:
Adding Destroy APIs and sample call

essamROCm committed:
Adding rocJPEG APIs & calling from the sample

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG

essamROCm committed:
Adding more rocJPEG APIs and their pybinding plus sample code

essamROCm committed:
Adding HIP Alloc-Free-Check APIs

essamROCm committed:
Adding more rocJPEG APIs

essamROCm committed:
Py Sample complete, with one sig fault in hip free

essamROCm committed:
Correcting widths and heights to be an array

essamROCm committed:
Moving all storage-variables to user space out of the main class

essamROCm committed:
Splitting Utils class out and adding hip python Alloc-Free calls

essamROCm committed:
Clean up variables names, extra empty lines

essamROCm committed:
Two methods to call HIP

essamROCm committed:
Return all changed values to the user to re-use

essamROCm committed:
Moving PyInitDecodeParams from decode to Utils class

essamROCm committed:
Making sure address are passed correctly nor issues, removing un-needed code

essamROCm committed:
Clean up sample, change ptr type attempting to fix sig fault

essamROCm committed:
wrapping void * in a safe capsules, destroy APIs implied in the capsules, preparing batch API

essamROCm committed:
adding debug code, making sure alloc ret an address

essamROCm committed:
testing with original sample code

essamROCm committed:
Adding Batched Jpeg Decode Sample and related APIs

essamROCm committed:
Fix for build error

essamROCm committed:
Adding missing variables to the sample removing debug code

essamROCm committed:
Rewrote the batch sample with roi modified its APIs

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG

essamROCm committed:
Jpeg class is a standalone class, Utils class not inheriting from Jpeg class

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG

essamROCm committed:
Removing un-necessary code and include headers

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJPEG

essamROCm committed:
Fix for correct type of parameter and init var variable in py sample

essamROCm committed:
Removing extra debug code from sample

essamROCm committed:
Remove debug code from batched sample, NO SIG FAULT in both batched and single image decode

essamROCm committed:
Correcting batch process logic, adding validation to params in samples

essamROCm committed:
Modifying code for review comments, CMakeLists.txt, remove .cmake file

essamROCm committed:
Proper error check examining every status return value in all calls

essamROCm committed:
All status check and error handling moved to sample, batch sample now identical to c++ sample, still sig fault issues to be resolved

essamROCm committed:
Logic fix was causing sig fault, now fixed

essamROCm committed:
batch py sample output is identical to c++ sample output, except in measured time

essamROCm committed:
Proper error handling in single image decoding python sample

essamROCm committed:
Adding rocjpeg as dependency, adding ctest for jpegdecode

essamROCm committed:
Adding batched sample in the ctest list

essamROCm committed:
Correcting if version in 2 Cmake, removing repeated lines

essamROCm committed:
Arguments redone in original order
renamed variables in API decoder
returned the 'Destroy' APIs back
renamed Cmake vars to have meaning
corrected dates in multiple files
added self assignment in the decoder API
checked crop rect at main

essamROCm committed:
Fix for jpegdecode loop logic, error messages

essamROCm committed:
Changes made for the requests in PR review conversation

essamROCm committed:
CMake fix, and adding full block of copyrights

essamROCm committed:
1st Draft API and sample as was discussed in team design review meeting

essamROCm committed:
decode single file, batch of files, byte-data or batch of byte-data, numpy array or batch of numpy arrays, all from the same interface same as we discussed in the design review meeting, no bugs

essamROCm committed:
Adding the missing DecodeSource() class sample call, and uncomment dlpack messages

essamROCm committed:
uncomment dlpack tensor, and its messages

essamROCm committed:
Remove old code, correct DecodeSource() call

essamROCm committed:
Reorganize tests & samples, create matching notebooks

essamROCm committed:
Removing un-used code, get CTest to work, rename tests to appropriate naming

essamROCm committed:
Removing un-used code, get CTest to work, rename tests to appropriate naming

essamROCm committed:
Saving latest notebooks changes

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJpegDecode_Python

essamROCm committed:
Optimizing sample notebook code

essamROCm committed:
Correcting switched test sample file names

kiritigowda committed:
Merge branch 'develop' into ea/Adding_rocPyJpegDecode_Python

essamROCm committed:
Fix batch decode bug, adding new batch decode sample

essamROCm committed:
Adding interleaved ROCJPEG_OUTPUT_RGB as default output image format

essamROCm commented:
Nothing more to add/modify or to correct for this draft PR, needs design review and approval, then to be moved to review-PR.
The overall design structure is implemented and tested thoroughly with 4 test python samples and 2 notebooks.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,59 @@
+    def __init__(
+            self,
+            param):
+        self.DS = rocpyjpeg.DecodeSource(param)
```
rename `DS` to `decode_source`

essamROCm replied:
Renamed.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,59 @@
+class decoder(object):
+    def __init__(
+            self, 
+            device_id = 0, 
```
save `device_id` for future use in dlpack. refer to the video decoder which is already added

essamROCm replied:
Saved.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,59 @@
+        self.DS = rocpyjpeg.DecodeSource(param)
+
+    # return the code stream
+    def cs(self):
```
rename `cs` to `code_stream`

essamROCm replied:
Changed to code_stream.

LakshmiKumar23 commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,130 @@
+{
```
add copyright for all notebooks

essamROCm replied:
Added.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,59 @@
+
+import rocpyjpegdecode as rocpyjpeg
+import rocpyjpegdecode.jpegTypes as jpegt
+import numpy as np
```
remove if not used

essamROCm replied:
Removed.

LakshmiKumar23 commented on pyRocVideoDecode/decoder.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
+    # Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
remove this

essamROCm replied:
Removed.

LakshmiKumar23 commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,130 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "device_id = 0\n",
```
add comments for changing `device_id` if needed, `rocm_path` etc

essamROCm replied:
Added comments.

LakshmiKumar23 commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,130 @@
+    "cs = jdec.DecodeSource(img_full_path).cs()\n",
+    "width = cs.width\n",
+    "height = cs.height\n",
+    "print(width, height)\n",
```
remove this print, prints out of order

essamROCm replied:
Done.

LakshmiKumar23 commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,130 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def display_tensor(tensor):\n",
```
redo this with `cv2.imshow()`

essamROCm replied:
I tried, it doesn't work, might need a lot of changes which complicate not simplify the function. CV2 doesn't work OK with notebooks, it is recommended to use PIL instead. It is just 2 lines to display the image; user can use different method(s).

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,275 @@
+    // Get Channel Pitch And Sizes
+    memset(&decode_params, 0, sizeof(RocJpegDecodeParams));
+    memset(&output_image, 0, sizeof(RocJpegImage));
+    decode_params.output_format = ROCJPEG_OUTPUT_RGB; // width will be * 3
```
Add an API for `SetOutputFormat` and set this based on that

essamROCm replied:
Implemented and used.

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,59 @@
+        return self.img        
+
+    # decode image or batch of images
+    def decode(self, img_full_bath):
```
typo - `path`

essamROCm replied:
Corrected.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,478 @@
+            std::cerr << "Warning: Null DecodeSource* skipped." << std::endl;
+            // push empty/invalid record
+            CodeStream empty_code_stream;
+            empty_code_stream.set_valid(false);
```
needs to be changed to reuse code_stream for only number of batch size, not all images

essamROCm replied:
Done and tested.

essamROCm committed:
The latest files batch gets deleted and de-allocated when new batch is decoded or when decoder class is destroyed. All team comments and desgin changes were applied and implemented.

essamROCm committed:
Removing debug printout, adding the API/call to set format

essamROCm committed:
Correct output in the notebook

essamROCm committed:
Removing unnecessary host memory allocation and copying using .cpu() suffice

essamROCm committed:
Remove debug statements printout and code

essamROCm committed:
Adding device id user arg to sample, remove un-needed code

LakshmiKumar23 commented on pyRocJpegDecode/decoder.py:
```diff
@@ -0,0 +1,66 @@
+            self, 
+            device_id = 0, 
+            backend = 0):
+        self.jpegdec = rocpyjpeg.Decoder(device_id, backend)
```
move this after initialization. Standard practice is to initialize it and pass as function arg.. It should be
```
def __init__(
            self, 
            device_id = 0, 
            backend = 0):
        self.device_id = device_id
        self.backend = backend
        self.jpegdec = rocpyjpeg.Decoder(self.device_id, self.backend)
```

essamROCm replied:
Done in last commit: https://github.com/ROCm/rocPyDecode/pull/178/commits/933269e3ba756a2492d647ef0d12daf363e65983

LakshmiKumar23 commented on samples/rocjpeg/batch_sample.ipynb:
```diff
@@ -0,0 +1,123 @@
+{
```
add copyright

essamROCm committed:
Adding the PLANAR to the dlpack wrapping with switch case

LakshmiKumar23 commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,170 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAEOCAIAAADe+FMwAAEAAElEQVR4nKT9SbNsW3IeiH3uvvaOiNPc/t3XZofMRBIJEAWwimRVSaYqaiQzDWSmmTTURFONpb+gicrU1FxWoiQzSRMWzSRTycQSiz0AkmAiCSRevsyXr7/9aSP2Xu6fBr7Wjn1ugjISijR7ee45ETv2Xo0v988//1w+/8/+CxEhKSKqCkBEyEABIKggqaokzczd850UCEFSiYjwomYWESKiwYiAR152+Xj+E0CN2J3sZgYAZUV/MQIi+bO0/5gIIAGQBEQoIFFKcfdaZ1UFJK9cRCLvQARgRJRSIoj2YYhIRABQFZGiau6TqtpQPCLfs7wEIiQAJ0s ...
```
not sure what all this is, maybe you ran the notebook then saved?

essamROCm replied:
Yes, I did this intentionally to match what 'others' do, so when you view our notebook online you can see the results of running each cell without actually running it. 

Should I save it clean without the image and the runtime results? Remember that we are representing what the customer wants to see.

essamROCm replied:
I can resave the notebook, please discuss with the team.

LakshmiKumar23 commented on samples/rocjpeg/test_batched_files_process.py:
```diff
@@ -0,0 +1,163 @@
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+def jpeg_decode(input_file_path, single, batch, output_format, device_id_in):
+
+    device_id = device_id_in    # example, change it to your system GPU index
```
not required, can be used directly

essamROCm replied:
I am planning to REMOVE all 'test' files .py, they were created just to test the implementation with debug messages. The end user shouldn't see them. The actual samples would be only the notebooks.

LakshmiKumar23 commented on samples/rocjpeg/test_batched_files_process.py:
```diff
@@ -0,0 +1,163 @@
+    print(f"Image output format set to: {jpegt.RocJpegOutputFormat(output_format)}")
+
+    line = '\n' + '-' * 60 + '\n'
+    root_folder = input_file_path   # user input
```
no need for these 2 lines

essamROCm replied:
I am planning to REMOVE all 'test' files .py, they were created just to test the implementation with debug messages. The end user shouldn't see them. The actual samples would be only the notebooks.

LakshmiKumar23 commented on samples/rocjpeg/test_batched_files_process.py:
```diff
@@ -0,0 +1,163 @@
+    root_folder = input_file_path   # user input
+    batch_size = batch              # user input
+
+    if(single >= 1):
```
we do not need this either. Take only `-b` as input. can be default 0. then do `decode_single`. If user passes `-b 1` or greater than 1, can do `decode_batch`. Can remove the `single` input and function arg

essamROCm replied:
I am planning to REMOVE all 'test' files .py, they were created just to test the implementation with debug messages. The end user shouldn't see them. The actual samples would be only the notebooks.

LakshmiKumar23 commented on samples/rocjpeg/test_batched_files_jpeg_decode.py:
```diff
@@ -0,0 +1,100 @@
+    arrays_list.append(np.fromfile(p, dtype=np.uint8))
+images_from_array_list = decoder.decode(arrays_list)
+for img in images_from_array_list:
+    printout_tensor_info(img)
```
this file seems to me like it only tests the function calls with different options of file/arrya vs data. I would suggest move this under `tests/` directory. I do not think this needs to be under samples

essamROCm replied:
I am planning to REMOVE all 'test' files .py, they were created just to test the implementation with debug messages. The end user shouldn't see them. The actual samples would be only the notebooks.

LakshmiKumar23 commented on samples/rocjpeg/test_decode_source.py:
```diff
@@ -0,0 +1,87 @@
+    print(f"\nimage [{i}] width: {width}\nimage[{i}] height: {height}")
+images = decoder.read(cs_list)
+for i, img in enumerate(images):
+    printout_tensor_info(img)
```
same as above, please move under `tests/`

essamROCm replied:
I am planning to REMOVE all 'test' files .py, they were created just to test the implementation with debug messages. The end user shouldn't see them. The actual samples would be only the notebooks.

LakshmiKumar23 commented on samples/rocjpeg/test_single_file_jpeg_decode.py:
```diff
@@ -0,0 +1,106 @@
+np_arr = np.fromfile(img_full_path, dtype=np.uint8)
+print(line, "TEST (3) array (numpy)", line)
+image = decoder.decode(np_arr)
+printout_tensor_info(image)
```
can combine this file with `test_batched_files_jpeg_decode` since it is testing the same `.decode` or `.read` functions. Rename combined file to `testjpeg_decode.py`

essamROCm replied:
I am planning to REMOVE all 'test' files .py, they were created just to test the implementation with debug messages. The end user shouldn't see them. The actual samples would be only the notebooks.

essamROCm committed:
Adding PLANAR sample making use of PLANAR output format, correcting PLAN dim

essamROCm committed:
Updating the rgb_planar.ipynb sample

essamROCm committed:
Using self init values in init call in decoder API

rrawther commented on src/rocjpeg/roc_pyjpeg.h:
```diff
@@ -0,0 +1,92 @@
+
+    // to export as GPU MEM (dlpack)
+    std::vector<std::shared_ptr<BufferInterface>> ext_buf;
+
```
add comment for this function

essamROCm replied:
Done

rrawther commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,286 @@
+    }
+    // Read file data if no data sent
+    std::vector<char> file_data;
+    int file_size = data_size;
```
you can use the data_size directly in line#151, get rid of this line

essamROCm replied:
Done

rrawther commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,286 @@
+    return EXIT_SUCCESS;
+}
+
+int CodeStream::clean_up_return_fail() {
```
I suggest renaming the function to something more abstract line "release()"

essamROCm replied:
Done

rrawther commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,286 @@
+    return *this;
+}
+
+int CodeStream::ReadImageFromDiskFile(const std::filesystem::path& filename, std::vector<char>& file_data, int& file_size) {
```
suggest renaming fun name "ReadFromFile()"

essamROCm replied:
Done

rrawther commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,286 @@
+}
+
+// Use the dat and its size if valid, otherwise use the file to load the data
+int CodeStream::PrepareStreamForOneImageDecoding(const std::filesystem::path& filename, const unsigned char* data, int data_size) {
```
Please use more condensed function name something like "InitializeSingleImage()"

essamROCm replied:
Done

rrawther commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,427 @@
+}
+
+Decoder::Decoder(int device_id, int backend, RocJpegOutputFormat output_format) {
+    set_output_format(output_format);   // save user choise if sent for output format
```
Is there a purpose for using both CamelCase and snake_case for fn names

essamROCm replied:
Yes, this is a constructor of a class named 'Decode' so the constructor must be the same name (CamelCase). The class is named that way to 'MATCH' what we reviewed in the 1st design meeting. While the function better to be snake case for it is used in python as API.

rrawther commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,427 @@
+    PY_CHECK_ROCJPEG(rocJpegCreate(m_backend, m_device_id, &rocjpeg_handle));
+}
+
+PyJpegImages Decoder::decode(DecodeSource* data) {
```
change to Camel case

essamROCm replied:
Here is not a constructor, it is actual function that performs the decode, also used in the python, so it is better to stay snake case. Please see my comment above about the constructor vs. py API function name.

rrawther commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,65 @@
+    std::vector<CodeStream> code_stream; // add all live instance
+
+    // STORE: to export IMAGE/IMAGES-BATCH to python
+    std::vector<PyJpegImages> image;                        // all live instance
```
suggest following google coding style for variable names. please rename to "images_" to represent vector

essamROCm replied:
Done

rrawther commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,427 @@
+            CodeStream empty_code_stream;
+            empty_code_stream.set_valid(false);
+            code_stream.push_back(empty_code_stream);
+            std::cout << "Added empty code stream" << std::endl;
```
please remove this comment if it is for debugging

essamROCm replied:
Done

rrawther commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,427 @@
+    // loop the whole list length
+    for (auto* data : decode_source_arg) {
+        if (!data) {
+            std::cerr << "Warning: Null DecodeSource* skipped." << std::endl;
```
Is this a valid case? Why are we creating an empty code stream?

essamROCm replied:
Yes, some files in the batch might be invalid or not supported, we have to keep 'invalid' record of it, same as the original design is doing. The user in the PY use cases checks if the image in the list valid to use or not before using.

rrawther commented on samples/rocjpeg/test_decode_source.py:
```diff
@@ -0,0 +1,87 @@
+device_id = 0
+backend = 0
+
+img_full_path = "/opt/rocm/share/rocjpeg/images/mug_420.jpg"
```
It is better to use environment variable {ROCM_PATH} instead of hardcoded path as the user can install rocm anywhere

essamROCm replied:
This test_..py and other 4 ones will be removed they were just for implementation testing, not for end user.

rrawther commented on samples/rocjpeg/test_single_file_jpeg_decode.py:
```diff
@@ -0,0 +1,106 @@
+device_id = 0
+backend = 0
+
+img_full_path = "/opt/rocm/share/rocjpeg/images/mug_420.jpg"
```
use environment for ROCM_PATH

essamROCm replied:
This test_..py and other 4 ones will be removed they were just for implementation testing, not for end user.

rrawther commented on samples/rocjpeg/batch_sample.ipynb:
```diff
@@ -0,0 +1,120 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)\n",
```
The font size is too big in the notebook display. Please reduce to appropriate size

rrawther commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,170 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeAAAAEOCAIAAADe+FMwAAEAAElEQVR4nKT9SbNsW3IeiH3uvvaOiNPc/t3XZofMRBIJEAWwimRVSaYqaiQzDWSmmTTURFONpb+gicrU1FxWoiQzSRMWzSRTycQSiz0AkmAiCSRevsyXr7/9aSP2Xu6fBr7Wjn1ugjISijR7ee45ETv2Xo0v988//1w+/8/+CxEhKSKqCkBEyEABIKggqaokzczd850UCEFSiYjwomYWESKiwYiAR152+Xj+E0CN2J3sZgYAZUV/MQIi+bO0/5gIIAGQBEQoIFFKcfdaZ1UFJK9cRCLvQARgRJRSIoj2YYhIRABQFZGiau6TqtpQPCLfs7wEIiQAJ0s ...
```
I think this should be cleared before pushing the file

essamROCm replied:
Done

rrawther commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,170 @@
+{
+ "cells": [
```
All jupyter notebook outputs needs to be cleared before pushing the file into the repo

essamROCm replied:
Done

rrawther commented on samples/rocjpeg/rgb_planar.ipynb:
```diff
@@ -0,0 +1,165 @@
+{
+ "cells": [
```
notebook outputs needs to be cleared before pushing it into the repo

essamROCm replied:
Done

rrawther reviewed (changes requested):
Added review comments

rrawther reviewed (changes requested):
Added some more comments

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg.h:
```diff
@@ -0,0 +1,92 @@
+
+public:
+    PyJpegImages() {
+        // init GPU mem -python buffer
```
add a comment to say we need this for Y, U and V planes

essamROCm replied:
Done

essamROCm replied:
src/rocjpeg/roc_pyjpeg.h **LINE: 43** comments added.

essamROCm committed:
Finalizing implementing all requestes, ctest working correct, adding 2 simple PY samples, removing test code.

essamROCm committed:
Adding final comment in header file for TODO: YUV image case

LakshmiKumar23 commented on samples/rocjpeg/decode_rgb_planar.ipynb:
```diff
@@ -0,0 +1,144 @@
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
output does not display correctly. Will share image offline

essamROCm replied:
Re-tested, it displays the intents for demonstration, which is the R PLANE of the output image as red, G plane as green, B plane as blue, then in the same row the whole image as RGB, which is by python code I collected and mixed the 3 planes.

LakshmiKumar23 commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,131 @@
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
incorrect output display

essamROCm replied:
Same as the PLANAR example I retested it ant shows the output image correctly, if it doesn't on the test machine we can look at it for further inspection.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,110 @@
+
+    if(img.is_valid()):
+        yuv_tensor = torch.from_dlpack(img.ext_buf[0].__dlpack__(img))
+        print(f"Device:\t\t\t {yuv_tensor.device}")
```
please remove prints, not required. Add option for `-o` to save the file instead.

essamROCm replied:
Removed.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,127 @@
+
+    print(f"Image output format set to: {jpegt.RocJpegOutputFormat(output_format)}")
+    print(f"\nDecoding whole folder of images: {input_file_path} on Device: {'CPU' if backend else 'GPU'} with Device ID: {device_id}")
+    decode_batch(decoder, input_file_path, batch_size)
```
why do we need a separate function? It can just be one function `jpeg_decode_batch`

essamROCm replied:
We need a separate function because rocJPEG has two distinct APIs, one for single file decode and different call with different parameters for batched image(s).

rrawther replied:
I think what Lakshmi is referring to is combining jpeg_decode and decode_batch into a single funcion called "jpeg_decode_batch" which makes sense

essamROCm replied:
I converted it to one function now.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,127 @@
+ 
+
+def decode_batch(decoder, folder_full_path, batch_size):
+    if not isinstance(batch_size, int) or batch_size <= 0:
```
this check can be done at arg parse level and exit immediately. No need to go through 2 functions

essamROCm replied:
The check moved to arg parse level, and it exits if incorrect.

LakshmiKumar23 commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,127 @@
+    if not isinstance(batch_size, int) or batch_size <= 0:
+        print(f"Args Error: batch_size must be a positive integer, got {batch_size}\n")
+        return
+    if not os.path.isdir(folder_full_path):
```
same with this parameter; combine with line 123

essamROCm replied:
This check too, moved to arg parse level, and it exits if incorrect.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_buffer.cpp:
```diff
@@ -0,0 +1,185 @@
+using namespace std;
+using namespace py::literals;
+
+static void CheckValidBuffer(const void *ptr) {
```
why are we redefining functions that are already available in src?
[Link](https://github.com/ROCm/rocPyDecode/blob/aa3be12fae277331d400dfe96fd462cdc175a2e9/src/roc_pybuffer.cpp#L32) Please use common functions for both videoDecode and jpegDecode. The dlpack/buffer functions need not be rewritten

LakshmiKumar23 replied:
applies to the whole file it seems like. Please remove this

essamROCm replied:
Implemented in: https://github.com/ROCm/rocPyDecode/pull/178/commits/48194bc31ee2c363185a068abb82e98554ff1779

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_buffer.cpp:
```diff
@@ -0,0 +1,185 @@
+    }
+}
+
+BufferInterface::BufferInterface(DLPackPyTensor &&dlTensor) {
```
same as above

essamROCm replied:
Implemented in: https://github.com/ROCm/rocPyDecode/pull/178/commits/48194bc31ee2c363185a068abb82e98554ff1779

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_buffer.cpp:
```diff
@@ -0,0 +1,185 @@
+
+const DLTensor &BufferInterface::dlTensor() const {
+    return *m_dlTensor;
+}
```
same as above for all functions till here

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_buffer.h:
```diff
@@ -0,0 +1,63 @@
+/*
```
file repeated from existing src/

essamROCm replied:
Implemented in: https://github.com/ROCm/rocPyDecode/pull/178/commits/48194bc31ee2c363185a068abb82e98554ff1779

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+    std::copy(std::begin(other.heights), std::end(other.heights), std::begin(heights));
+    std::copy(std::begin(other.channel_sizes), std::end(other.channel_sizes), std::begin(channel_sizes));
+    num_channels = other.num_channels;
+    code_stream_ = other.code_stream_;  // raw pointer copy (or convert to shared_ptr if needed)
```
please use `<shared_ptr>` as it maintains reference counts and hence avoids memory leaks

essamROCm replied:
Changed to <shared_ptr> in roc_pyjpeg_codestream.h **LINE: 52**

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+    }
+    // Read file data if no data sent
+    std::vector<char> file_data;
+    int file_size = data_size;
```
remove this line since `data_size` is directly used

essamROCm replied:
LINE: 143 removed, variable -reused.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+        file_data.resize(data_size);
+        file_data.assign(reinterpret_cast<const char*>(data), reinterpret_cast<const char*>(data) + data_size);
+    }    
+    if(data == nullptr) {
```
can be else statement

essamROCm replied:
Changed to 'else' statement.

AryanSalmanpour commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,110 @@
+    img = decoder.decode(input_file_path)
+
+    if(img.is_valid()):
+        yuv_tensor = torch.from_dlpack(img.ext_buf[0].__dlpack__(img))
```
The name `yuv_tensor` is incorrect because the default output format is `ROCJPEG_OUTPUT_RGB`. Please rename `yuv_tensor` to `tensor`.

essamROCm replied:
Removed from the sample now.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg.cpp:
```diff
@@ -0,0 +1,103 @@
+
+using namespace std;
+
+RocJpegHandle rocjpeg_handle = nullptr;     // main session
```
Why is rocjpeg_handle defined as a global variable here?

essamROCm replied:
Was experimental, now is not global.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+
+    // Create Stream
+    RocJpegStatus rocjpeg_status = ROCJPEG_STATUS_NOT_INITIALIZED;
+    rocjpeg_status = rocJpegStreamCreate(&stream_handle);
```
How many stream handles are created? If there is only one and it is being reused, why is there no check to determine if the stream handle was previously created?

essamROCm replied:
Added the check/de-alloc, although it is per-image, means only called one time via pybind/python to instantiate.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+CodeStream::CodeStream(const std::filesystem::path& filename) {
+    PY_CHECK_DECODER();
+    set_valid(false);
+    if( InitializeSingleImage(filename, nullptr, 0) == EXIT_SUCCESS) {
```
remove extra space: `if( InitializeSingleImage` -> `if (InitializeSingleImage`

essamROCm replied:
Removed.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+CodeStream::CodeStream(const unsigned char* data, size_t length) {
+    PY_CHECK_DECODER();
+    set_valid(false);
+    if( InitializeSingleImage(static_cast<const std::filesystem::path>(""), data, length) == EXIT_SUCCESS) {
```
remove extra space: if( InitializeSingleImage -> if (InitializeSingleImage

essamROCm replied:
Removed.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+    auto data_view = static_cast<std::string_view>(data_ref_bytes_);
+    py::gil_scoped_release release;
+    set_valid(false);
+    if( InitializeSingleImage(static_cast<const std::filesystem::path>(""), reinterpret_cast<const unsigned char*>(data_view.data()), data_view.size()) == EXIT_SUCCESS) {
```
remove extra space: if( InitializeSingleImage -> if (InitializeSingleImage

essamROCm replied:
Removed.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+    auto data = data_ref_arr_.unchecked<1>();
+    py::gil_scoped_release release;
+    set_valid(false);
+    if( InitializeSingleImage(static_cast<const std::filesystem::path>(""), data.data(0), data.size()) == EXIT_SUCCESS) {
```
remove extra space: if( InitializeSingleImage -> if (InitializeSingleImage

essamROCm replied:
Removed.

rrawther reviewed (commented):
I don't see commits reflecting the changes regarding review comments.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decode_source.h:
```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <filesystem>
+// namespace fs = std::filesystem;
```
remove this header if not used

essamROCm replied:
Removed.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decode_source.cpp:
```diff
@@ -0,0 +1,80 @@
+THE SOFTWARE.
+*/
+
+#include <filesystem>
```
remove since not used

essamROCm replied:
Removed.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,64 @@
+    static void exportToPython(py::module& m);
+
+    // reset what was allocated and filled before
+    void reset_code_stream_store();
```
`reset_code_stream()` and `reset_image()` seem like better names to me. is this just to say it stores the corresponding info?

essamROCm replied:
Renamed as suggested.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,64 @@
+    int m_device_id;
+    RocJpegBackend m_backend;
+    bool to_dlpack_tensor(CodeStream* code_stream, PyJpegImages* image);
+    bool get_widths_heights_from_output_format(std::vector<uint32_t>& widths, std::vector<uint32_t>& heights, uint32_t img_width, uint32_t img_height, RocJpegOutputFormat output_format, RocJpegChromaSubsampling subsampling);
```
rename to `get_output_dims` or something more concise

essamROCm replied:
Reamed.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,429 @@
+            }, "Get the shape of the U plane buffer as an array")
+        .def_property_readonly("shapeU", [](std::shared_ptr<PyJpegImages>& self) {
+            return self->ext_buf[1]->shape();
+            }, "Get the shape of the U plane buffer as an array")
```
I think this is V plane

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,429 @@
+            }, "Get the shape of the Y plane buffer as an array")
+        .def_property_readonly("shapeUV", [](std::shared_ptr<PyJpegImages>& self) {
+            return self->ext_buf[1]->shape();
+            }, "Get the shape of the U plane buffer as an array")
```
do you mean UV plane?

essamROCm replied:
Yes, same as in rocPyDecode will be used in YUV out format.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_dlpack.h:
```diff
@@ -0,0 +1,53 @@
+THE SOFTWARE.
+*/
+
+#ifndef USE_DLPACK_HEADER
```
same file as https://github.com/ROCm/rocPyDecode/blob/develop/src/roc_pydlpack.h
Please reuse and do not copy it here

essamROCm replied:
Implemented in: https://github.com/ROCm/rocPyDecode/pull/178/commits/48194bc31ee2c363185a068abb82e98554ff1779

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_dlpack.cpp:
```diff
@@ -0,0 +1,114 @@
+THE SOFTWARE.
+*/
+
+#include "roc_pyjpeg.h"
```
same file as https://github.com/ROCm/rocPyDecode/blob/develop/src/roc_pydlpack.cpp
Please reuse and do not copy it here

essamROCm replied:
Implemented in: https://github.com/ROCm/rocPyDecode/pull/178/commits/48194bc31ee2c363185a068abb82e98554ff1779

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg.cpp:
```diff
@@ -0,0 +1,103 @@
+
+        rocPyJpegDecode Python API reference
+
+        This is the Python API reference for the AMD ROCm rocJEPEG library.
```
typo `rocJPEG`

essamROCm replied:
Same as rocDecode, not typo meant to be rocJPEG.

LakshmiKumar23 replied:
rocJEPEG library -- on line 41. this is what i was pointing to. Similarly, I notice a lot of other typos across the PR. Please take another look and clean it up

essamROCm replied:
Corrected in last commit.

essamROCm committed:
Code-Changes for the team review requests

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,284 @@
+int CodeStream::release() {
+    hipError_t hip_status = hipSuccess;
+    if(stream_handle) {
+        RocJpegStatus rocjpeg_status = rocJpegStreamDestroy(stream_handle);
```
Please set `stream_handle` to `nullptr ` after destroying it.

essamROCm replied:
stream_handle set to **nullptr** after dealloc.

essamROCm committed:
setting stream_handle to nullptr, correct input check in sample, add cerr msg

essamROCm committed:
- Created new folder under src/: common, and rocdecode
- Moved rocPyDecode source files into src/rocdecode
- Kept rocPyJpeg source files into src/rocjpeg without the buffer/dlpack files
- Moved the dlpack and buffer CPP & H files into src/common new folder
- All code that refers to buffer/dlpack was modified to share the files from src/common
- Created new samples/rocdecode sub-folder
- Moved rocPyDecode Python Samples under samples/rocdecode
- Modified CMake files to find the samples when installing
- Modify CMake for ctest to find the samples
- Added vector to keep single decoded image up to MAX of 16
- Displaying error msg of exceeding the max advised to use batched instead
- Removing "keep_alive" from pybind, as we pass only ONE parameter

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,20 @@
+
+### Added
+
+* rocpyjpegdecode package 
```
This needs more context. Is this for installation? How does it affect the user?

essamROCm replied:
It will affect the user who already used rocPyDecode before specifically in finding where the samples are. We will need some changes in the documentation to reflect moving the samples under subfolder. We are talking about 6.5 and above.

spolifroni-amd replied:
OK in that case, the info in the changelog needs to be enhanced to explain this.

essamROCm replied:
Added what have changed of new subfolder to the CHANGELOG.md

spolifroni-amd replied:
Thanks Essam! This is a lot better and a lot more useful!

LakshmiKumar23 commented on src/common/roc_pybuffer.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
+Copyright (c) 2025 Advanced Micro Devices, Inc. All rights reserved.
```
change the old files to `2024 - 2025`

essamROCm replied:
Changed.

LakshmiKumar23 commented on src/common/roc_pybuffer.cpp:
```diff
@@ -20,6 +20,7 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */
 
+// #include "roc_pyjpeg.h"
```
remove unnecessary comments

essamROCm replied:
Removed.

LakshmiKumar23 commented on src/common/roc_pybuffer.h:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
+Copyright (c) 2025 Advanced Micro Devices, Inc. All rights reserved.
```
`2024-2025`

essamROCm replied:
Changed.

LakshmiKumar23 commented on src/common/roc_pydlpack.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
+Copyright (c) 2025 Advanced Micro Devices, Inc. All rights reserved.
```
same as above

essamROCm replied:
Changed.

LakshmiKumar23 commented on src/common/roc_pydlpack.cpp:
```diff
@@ -20,14 +20,15 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 */
 
-//#include <pybind11/stl.h>
+// #include "roc_pyjpeg.h"
```
remove comment

essamROCm replied:
Removed.

LakshmiKumar23 commented on src/common/roc_pydlpack.cpp:
```diff
@@ -20,14 +20,15 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 
-//#include <pybind11/stl.h>
+// #include "roc_pyjpeg.h"
+#include <pybind11/stl.h>
```
remove if not used

essamROCm replied:
Removed.

LakshmiKumar23 commented on src/common/roc_pydlpack.h:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
+Copyright (c) 2025 Advanced Micro Devices, Inc. All rights reserved.
```
same as above

essamROCm replied:
Changed.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,443 @@
+    // clang-format on
+}
+
+RocJpegHandle Decoder::rocjpeg_handle = nullptr;                         // main session
```
why is this handle and format global?

rrawther replied:
I agree. We can't really use global sessions here as multi-instances of this can cause issues

essamROCm replied:
This is not global, it is contained inside the Decoder Class, and only carry one instance info, means if you run multi-instances for decoding a new instance of those associated with the new Decoder class are created and contained in the new class. We can discuss further if you like, also a test code can show better. Those 2 vars are declared static too to be contained and protected by the class instance.

rrawther replied:
Initialize it in the class definition itself

essamROCm replied:
I tried but for c++ rules it must be done outside the class definition.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,76 @@
+private:
+    int m_device_id;
+    RocJpegBackend m_backend;
+    int m_batch_size = 2; // default, not used for now
```
default can be 0

essamROCm replied:
This is removed, not used or needed.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,76 @@
+    void set_output_format(RocJpegOutputFormat output_format);
+
+    // STORE: keep code_stream(s) info here
+    std::vector<CodeStream> code_stream;        // one batch instances
```
why do we need two different ones. I would suggest use the same. Only one case will be used at a time, either batched or single decoding. Default batch_size can be 0. If 0 treat it as single image decoder. If user sets batch size use it for batchedDecoding.
@rrawther  @AryanSalmanpour  can chime in too

essamROCm replied:
We need 2 different storages here, one for singles that can grow up to MAX=16 as we discussed in the last meeting, while the other storage can be reset for every new batch or when decoder class is destroyed. other things can be done but complicating the code for no reason, one or two stores will eventually store the same size, however with 2 the latter one can be freed freely without complication in the code.

rrawther replied:
Yes. I think we should have only one. If it is single image, the codestream vector contain only one instance. Same for line#57

essamROCm replied:
The code_stream vector will contain more than one instance, because we agreed to keep what the user decoding up to MAX count like 16, 24 as we discussed in the meeting. That is why there are 2 stores, to be easy to RESET the one that saves the batches.

rrawther commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,287 @@
+CodeStream::CodeStream() {
+}
+
+CodeStream::~CodeStream() {
```
don't you want to check here for sanity, if stream_handle is released. User might create a codestream and not call release

essamROCm replied:
I did during debugging, no need at runtime. Pybind11 creates an instance and later after we consume it gets safely destroyed.

rrawther commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,287 @@
+    return EXIT_SUCCESS;
+}
+
+int CodeStream::release() {
```
change to Release() so it follows the same coding style as other functions in this file

essamROCm replied:
Changed.

rrawther commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,76 @@
+    int m_device_id;
+    RocJpegBackend m_backend;
+    int m_batch_size = 2; // default, not used for now
+    static RocJpegHandle rocjpeg_handle;     // main session
```
why these variables need to be static?

essamROCm replied:
To be contained and protected by the class instance. This adding to my comment above that those are not considered global variables.

rrawther commented on src/rocjpeg/roc_pyjpeg_utils.h:
```diff
@@ -0,0 +1,249 @@
+     * @param alignment The alignment value.
+     * @return The aligned value.
+     */
+    static inline int align(int value, int alignment) {
```
I think this can be moved out of the class.

essamROCm replied:
It is used extensively inside the PyRocJpegUtils, and not used anywhere outside.

rrawther reviewed (changes requested):
Added more review comments

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,443 @@
+
+void Decoder::reset_images(std::vector<PyJpegImages>& imgs) {
+    if (!imgs.empty()) {
+        // std::cout << "Cleaning up batch image store (" << image.size() << " instances.)" << std::endl;
```
remove

essamROCm replied:
Removed.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,443 @@
+        for (auto& img : imgs) {
+            if(img.is_valid()) {
+                for (auto& buf : img.ext_buf) {
+                    buf = std::make_shared<BufferInterface>();
```
why are we creating new pointers for `reset_images` . If `imgs` existed, they should have a buffer associated with it already. We should reuse them

essamROCm replied:
I did not get your question. Can you explain more please. The image class is instantiated for every new instance of the code_stream which is done by pybind, a new image must be initialized.

essamROCm committed:
Fix for GPU-Hang bug, Changing the code to all of the code viewer requestes.

rrawther commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,431 @@
+    const CodeStream& c_stream = *data->code_stream();
+    code_stream_single.push_back(c_stream);
+    // runtime sanity check
+    if(!code_stream_single.back().is_valid()) {
```
please get rid of code_stream_single and use the same code_stream vector for both single and batch decoding

essamROCm replied:
In the meeting we agreed to keep what the user decoding as single image up to a MAX we specify. So, you do not want to do that anymore? If you want to keep that mechanism then 2 stores are appropriate, if you do not want to keep what the user decoding up to MAX images, then I can return to use one store. Using 2 makes it simpler in coding. Please advise.

essamROCm replied:
Changes now made to get rid completely of any storage of decoding instances.

rrawther commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+private:
+    int m_device_id;
+    RocJpegBackend m_backend;
+    static RocJpegHandle rocjpeg_handle;     // main session
```
why is this static

essamROCm replied:
To be contained and protected by the class instance. This adding to my comment above that those are not considered global variables.

essamROCm committed:
Merging 2 func into one in batched sample

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+    static void exportToPython(py::module& m);
+
+    // reset what was allocated and filled before
+    void reset_code_streams(std::vector<CodeStream>& cs);
```
change `reset_code_streams` -> `ResetCodeStreams`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+
+    // reset what was allocated and filled before
+    void reset_code_streams(std::vector<CodeStream>& cs);
+    void reset_images(std::vector<PyJpegImages>& imgs);
```
change `reset_images` -> `ResetImages`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+    void reset_images(std::vector<PyJpegImages>& imgs);
+
+    // set batch size for batch process
+    void set_output_format(RocJpegOutputFormat output_format);
```
change `set_output_format` -> `SetOutputFormat`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+    std::vector<PyJpegImages> images_;          // one batch instance
+    std::vector<PyJpegImages> images_single;    // for single decode instances (up to MAX_SINGLE_DECODE)
+
+    static RocJpegHandle get_handle() {return rocjpeg_handle;};
```
change `get_handle` -> `GetHandle`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+    std::vector<PyJpegImages> images_single;    // for single decode instances (up to MAX_SINGLE_DECODE)
+
+    static RocJpegHandle get_handle() {return rocjpeg_handle;};
+    static void set_handle(RocJpegHandle h) { rocjpeg_handle = h;};
```
change `set_handle` -> ` SetHandle`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+
+    static RocJpegHandle get_handle() {return rocjpeg_handle;};
+    static void set_handle(RocJpegHandle h) { rocjpeg_handle = h;};
+    static RocJpegOutputFormat get_format() {return user_output_format;};
```
change `get_format` -> `GetFormat`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+    static RocJpegHandle get_handle() {return rocjpeg_handle;};
+    static void set_handle(RocJpegHandle h) { rocjpeg_handle = h;};
+    static RocJpegOutputFormat get_format() {return user_output_format;};
+    static void set_format(RocJpegOutputFormat fmt) { user_output_format = fmt;};
```
change `set_format` -> `SetFormat`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+private:
+    int m_device_id;
+    RocJpegBackend m_backend;
+    static RocJpegHandle rocjpeg_handle;     // main session
```
This rocjpeg_handle variable shouldn't be static. What issue do you have if you don't define rocjpeg_handle as static?

essamROCm replied:
Because other class CodeStream need to access its value, to avoid using global variables it is declared as static and set by the Decoder class, so it is not global, contained and protected by the Class Decoder.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+    static RocJpegHandle rocjpeg_handle;     // main session
+    static RocJpegOutputFormat user_output_format;    // dynamically adjusted by the user
+
+    bool to_dlpack_tensor(CodeStream* code_stream, PyJpegImages* image);
```
change `to_dlpack_tensor` -> `ToDlpackTensor`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,75 @@
+    static RocJpegOutputFormat user_output_format;    // dynamically adjusted by the user
+
+    bool to_dlpack_tensor(CodeStream* code_stream, PyJpegImages* image);
+    bool get_output_dims(std::vector<uint32_t>& widths, std::vector<uint32_t>& heights, uint32_t img_width, uint32_t img_height, RocJpegOutputFormat output_format, RocJpegChromaSubsampling subsampling);
```
change `get_output_dims `-> `GetOutputDims`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decode_source.h:
```diff
@@ -0,0 +1,50 @@
+    DecodeSource(const DecodeSource&) = delete;
+    DecodeSource& operator=(DecodeSource const&) = delete;
+
+    const CodeStream* code_stream() const;
```
change `code_stream()` -> `CodeStream()`

essamROCm replied:
Changed to CamelCase and HAD to make it **CodeStreamInstance**() because just CodeStream would collide with the class name and make tons of build issues.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_codestream.h:
```diff
@@ -0,0 +1,76 @@
+    uint32_t channel_sizes[ROCJPEG_MAX_COMPONENT] = {};
+    uint32_t num_channels = 0;
+    void set_valid(bool state) {valid = state;};
+    bool is_valid() {return valid;};
```
change `is_valid()` -> `IsValid()`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_codestream.h:
```diff
@@ -0,0 +1,76 @@
+    uint32_t heights[ROCJPEG_MAX_COMPONENT] = {};    
+    uint32_t channel_sizes[ROCJPEG_MAX_COMPONENT] = {};
+    uint32_t num_channels = 0;
+    void set_valid(bool state) {valid = state;};
```
change `set_valid` -> `SetValid`

essamROCm replied:
Changed to CamelCase.

AryanSalmanpour reviewed (commented):
@essamROCm I noticed additional formatting issues. Please use CamelCase for function names in the C++ classes. You mixed CamelCase (e.g., exportToPython) with lower_case (e.g., reset_code_streams, reset_images). It is important to maintain consistent naming conventions across the C++ classes.

essamROCm committed:
Added new 'save image' option in sample, and modification for the code review

essamROCm committed:
More CamelCase correction, and typo corrections

essamROCm commented:
I went through all the source checking for typos, I think I corrected all of them, all of what I was able to find, if you still find any that I missed please point to me the location of that typo to change it in future commits. Thank you.

essamROCm committed:
Changing the classes desgin based on team conversation

essamROCm committed:
Removing valid check not needed extra code

essamROCm committed:
Fix typos, CamelCases, moved notebook

essamROCm committed:
Merge branch 'develop' into ea/Adding_rocPyJpegDecode_Python

AryanSalmanpour commented on src/rocjpeg/roc_pyjpeg_decoder.h:
```diff
@@ -0,0 +1,56 @@
+    ~Decoder();
+
+    PyJpegImages decode(DecodeSource* data);
+    std::vector<PyJpegImages> decode(std::vector<DecodeSource*>& data_list);
```
Can you use a vector of `CodeStream` Instead of passing a vector of `DecodeSource`?

essamROCm replied:
No, because it is overloading to become code_stream at runtime (in the pybind declaration), that is the base of the design. The list arrives correct as code_stream objects.

AryanSalmanpour replied:
Let's discuss offline

essamROCm replied:
I am available online on team, let's discuss it.

LakshmiKumar23 commented on CHANGELOG.md:
```diff
@@ -2,6 +2,26 @@
+
+* rocpyjpegdecode package
+* Added src/rocjpeg source new subfolder
+* Moved rocPyDecode source files into new subfolder src/rocdecode
```
remove these points as the PR was separated

essamROCm replied:
Points removed.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,263 @@
+    // we return a list of images
+    std::vector<PyJpegImages> images_;
+
+    if(batch_size<=0)
```
formatting: add spaces `(batch_size <= 0)`

essamROCm replied:
Formatted.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,263 @@
+        // here 'images_' vector carries the count of 'valid' images
+        // to export to python (use dlpack(GPU MEM) {and numpy host array})
+        if (status == ROCJPEG_STATUS_SUCCESS) {
+            for(int i=0; i<count_of_valid_instances; i++) {
```
formatting: `for(int i = 0; i < count_of_valid_instances; i++)`

essamROCm replied:
Formatted.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_decoder.cpp:
```diff
@@ -0,0 +1,263 @@
+std::vector<PyJpegImages> Decoder::decode(std::vector<DecodeSource*>& decode_source_arg) {
+
+    RocJpegStatus status = ROCJPEG_STATUS_SUCCESS;
+    int batch_size = decode_source_arg.size();
```
where do we get this `decode_source_arg.size()`. Is this from the user or the number of files in the directory?

essamROCm replied:
Yes, from the user job he sends. This is the 'batch' size. If you have 1000 files but you send batch_size=3 that is the size of the 'decode_source_arg'.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_images.cpp:
```diff
@@ -0,0 +1,223 @@
+    switch(output_format) {
+        case ROCJPEG_OUTPUT_RGB_PLANAR: { // each color plane in a channel separately R[0], G[1], and B[2]
+            uint32_t surf_stride[3] = {widths[0], widths[1], widths[2]}; // ROCJPEG_OUTPUT_RGB_PLANAR all same width = img_width
+            for(int i=0; i<3; i++) {
```
formatting, add spaces `for(int i = 0; i < 3; i++)`

essamROCm replied:
Formatted.

essamROCm committed:
Changing code for review request, mostly formatting, remove changelog points, adding new one.

kiritigowda committed:
Merge branch 'develop' into ea/Adding_rocPyJpegDecode_Python

rrawther commented on samples/rocjpeg/decode_rgb_planar.ipynb:
```diff
@@ -0,0 +1,144 @@
+    "##### SPDX-License-Identifier:  MIT License\n",
+    "#### Python Sample - RGB Planar image"
+   ]
+  },
```
Add a description box saying what this sample is doing

essamROCm replied:
Added.

essamROCm committed:
Adding test calls to boost code coverage

essamROCm committed:
Adding description for what the planar notebook sample is doing.

rrawther commented on samples/rocjpeg/batch_sample.ipynb:
```diff
@@ -0,0 +1,118 @@
+    "##### Copyright © Advanced Micro Devices, Inc., or its affiliates.\n",
+    "##### SPDX-License-Identifier:  MIT License\n",
+    "#### Python Sample - Batch decode"
+   ]
```
Add short description block on what this sample does

essamROCm replied:
Added.

rrawther commented on samples/rocjpeg/decode_source.ipynb:
```diff
@@ -0,0 +1,131 @@
+    "### rocPyJpegDecode\n",
+    "##### Copyright © Advanced Micro Devices, Inc., or its affiliates.\n",
+    "##### SPDX-License-Identifier:  MIT License\n",
+    "#### Python Sample - Single file, data"
```
Add a short description block on what this sample does

essamROCm replied:
Added.

rrawther commented on samples/rocjpeg/jpegdecode.py:
```diff
@@ -0,0 +1,118 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
```
what is the purpose of this file since we already have notebook for the same?

essamROCm replied:
CTest, mainly to run the test, and also allow the user to pass parameters, the code is different.

rrawther commented on samples/rocjpeg/jpegdecodebatched.py:
```diff
@@ -0,0 +1,122 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import pyRocJpegDecode.decoder as jdec
```
what is the purpose of this file since we already have notebook for the same?

essamROCm replied:
CTest, mainly to run the test, and also allow the user to pass parameters, the code is different. 

Also here we pass a folder path, and take care of split it to batches, different example.

LakshmiKumar23 commented on CMakeLists.txt:
```diff
@@ -349,10 +393,18 @@ if(BUILD_ROCPYDECODE)
 
+    # Test - Installed -rocpyjpegdecode
+    install(DIRECTORY ${CMAKE_SOURCE_DIR}/tests DESTINATION ${CMAKE_INSTALL_DATADIR}/rocpyjpegdecode COMPONENT test)
+    install(DIRECTORY ${CMAKE_SOURCE_DIR}/cmake DESTINATION ${CMAKE_INSTALL_DATADIR}/rocpyjpegdecode COMPONENT test)
```
why are we installing `cmake` directory into tests?

essamROCm replied:
@kiritigowda 
Kiriti, can you please answer this question. That was your code during latest CMake changes.

essamROCm replied:
The answer from Kiriti is "Yes" this line is necessary we need to copy the cmake sub-folder into the tests folder.

essamROCm committed:
Adding short description to the notbook samples

rrawther reviewed (changes requested):
Added few minor comments

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_test.h:
```diff
@@ -0,0 +1,47 @@
+
+// Add all tests necessary to utilize 100% of the written code
+
+class PyJpegTest {
```
not sure if this should be placed under source, since these are test utilities

essamROCm replied:
I removed the code as it is proven to be ineffective, will discuss off-line a better method to boost code coverage.

LakshmiKumar23 commented on src/rocjpeg/roc_pyjpeg_test.cpp:
```diff
@@ -0,0 +1,115 @@
+THE SOFTWARE.
+*/
+
+#include "common/roc_pybuffer.h"
```
don't think this should be under `src`.
Seems like test utilities

essamROCm replied:
I removed the code as it is proven to be ineffective, will discuss off-line a better method to boost code coverage.

LakshmiKumar23 commented on samples/rocjpeg/jpegtest.py:
```diff
@@ -0,0 +1,25 @@
+
+import rocpyjpegdecode as rocpyjpeg
+
+test = rocpyjpeg.PyJpegTest()
```
let's discuss offline, if this needs to be part of the main library

essamROCm replied:
I removed the code as it is proven to be ineffective, will discuss off-line a better method to boost code coverage.

essamROCm committed:
Removing and Cleaning up ineffective test code that had been added solely to boost code coverage.

LakshmiKumar23 reviewed (approved):
Can merge if CI passes

essamROCm committed:
Ensure HIP is properly initialized; exit on failure if an invalid device ID is provided.

essamROCm committed:
Resolving copilot suggestions and comments

essamROCm commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,165 @@
+    // Stream Parse
+    rocjpeg_status = rocJpegStreamParse(reinterpret_cast<uint8_t*>(file_data->data()), data_size, stream_handle);
+    if (rocjpeg_status != ROCJPEG_STATUS_SUCCESS) {
+        std::cerr << "ERROR: Failed to parse the input jpeg stream with " << rocJpegGetErrorName(rocjpeg_status) << "Input File : " << (!filename.empty() ? filename : "") << std::endl;
```
Suggestion is applied in code now.

essamROCm commented on src/rocjpeg/roc_pyjpeg_codestream.cpp:
```diff
@@ -0,0 +1,165 @@
+        RocJpegStatus rocjpeg_status = rocJpegStreamDestroy(stream_handle);
+        stream_handle = nullptr;
+    }
+    return EXIT_FAILURE;
```
Code changed to reflect the suggestion.
